### ISSUE TITLE
Line Protocol should include unit of timestamp, precision should always truncate

### ISSUE THREAD

**Issue by sparrc** [influxdata/influxdb#4226]
As of now, writing to InfluxDB with the `precision` flag in the query string is inconsistent. When writing without a timestamp, the precision flag tells Influx to truncate the write time to the given unit of precision. When writing _with_ a timestamp, the precision flag is telling Influx the _units_ of the write.

Ergo, there is no way to give InfluxDB a timestamp in nanoseconds and tell it to write at second precision. The user must truncate their timestamp to seconds and then use the precision flag to tell Influx it's units.

Changing this will make the Influx write interface more consistent, and has the added benefit of allowing users to write using timestamps in any unit of measurement via other avenues, such as UDP.

If units are not given, it would default to nanoseconds, in the interest of backwards-compatability.

There is one use-case for which this will be a breaking change. Users who are writing via the HTTP interface with a timestamp will need to change the following line-protocol / query string combination:

query: `POST /write?precision=s`
line-protocol: `net_bytes_recv,host=tyrion,interface=en0 value=17412331189i 1000000000`

to

query: `POST /write?`
line-protocol: `net_bytes_recv,host=tyrion,interface=en0 value=17412331189i 1000000000s`

**tldr**

Allow the InfluxDB Line Protocol to include a timestamp (default to `ns`),

writes a value of 10 at time: 1000000000s
`cpu,host=foo value=10 1000000000s`

writes a value of 10 at time: 1000000000000000ns
`cpu,host=foo value=10 1000000000000000`

**Comment by corylanou** [comment]
So to be clear, the `s` parameter on the query will now be ONLY for truncating time, and unless specified in the line protocol, it's always `ns` time right?

**Comment by pauldix** [comment]
I think we should keep the query parameter as well. That way it's not a breaking change in any way. What's the advantage for making the query parameter not specify units as well as truncation?

**Comment by sparrc** [comment]
@corylanou correct, the `precision=` parameter in the query string would ONLY be for truncating time. Line protocol would default to `ns` unless specified otherwise

@pauldix Definitely keeping the query parameter, but the problem now is that it truncates in one situation (no timestamp), and it specifies units in another situation (with a timestamp). Doing both wouldn't really have any effect, because if you are specifying a timestamp in `s`, there is no truncation to do to get it to `s` anyways

**Comment by pauldix** [comment]
My suggestion is that if the units are specified on the line protocol, you ignore the query parameter. Otherwise, keep the same behavior we have now. That way we avoid the breaking change entirely.

**Comment by sparrc** [comment]
@pauldix then we won't be supporting truncating user-provided timestamps

**Comment by pauldix** [comment]
I don't think we need to support truncating user provided timestamps. That's trivial for them to do so there's no win from an API perspective.

**Comment by pauldix** [comment]
In fact, there's no reason for them to want truncation on a timestamp they provide. It's more efficient in terms of the amount of data you're sending over the wire to specify the precision and leave out all the excess digits

**Comment by corylanou** [comment]
So, the scenarios are:

## No timestamp provided
- precision parameter not provided: insert `now` with `ns` precision
- precision parameter provided: insert `now` with truncation to parameter specified precision

## Timestamp provided
- no precision parameter, no line protocol suffix: no truncation, time is considered `ns`
- precision parameter provided, no line protocol suffix: consider timestamp the parameter specified precision
- precision parameter provided, line protocol suffix provided: ignore param, use line protocol for timestamp precision, no truncation

Did I get that right?

**Comment by sparrc** [comment]
@corylanou yes that's correct, my original thought was for "timestamp provided, precision parameter provided, no line protocol suffix" to behave differently (truncating), but this way is certainly more backwards-compatible

**Comment by pauldix** [comment]
What @corylanou said is right. Assigned to you for 0.9.5 unless @sparrc wants to take it :)

**Comment by jsternberg** [comment]
Closing this in favor of #6037. I like the proposal here, but this is an old issue and it's going to get lost. This would also be a breaking change to the line protocol at this point and would have to go into a v2 anyway.

**Comment by pauldix** [comment]
@jsternberg just to confirm, you can still pass the unit in per point, but you're not required to?

**Comment by jsternberg** [comment]
In the current line protocol, I don't think you can. The current behavior is what's described at the beginning of this issue. I think we should move this discussion to #6037 because this would be a breaking change for the line protocol V1 and we don't want to forget about this proposal while considering what we should do for V2.
